### PR TITLE
UcrSave Column Position 

### LINKED ISSUE #3961: Position of calculated column in data frame
My particular example is the dialogue **Prepare > Column: Generate > Count in Factor**.  This puts the calculated column at the end of the data frame.  I would almost always prefer it just after the factor.
We have this option for **Prepare > Column: Calculate > Duplicate Column**.

So, for **Count in Factor**, perhaps we could make it simpler and give the option "**After Factor Column**" or "**At End**".  If we do this, then perhaps we could check whether this option could usefully be applied elsewhere too.

An alternative would be to have a simple dialogue/right-click (maybe just right-click) which makes an empty column and gives it an unusual name, e.g. "_no_name".  It contains NA throughout.  If this column exists, then it is automatically filled with a new calculated column. 

This is assigned to Danny/David first, because it needs a discussion.

### PR DESCRIPTION
**Pull request by Patowhiz**
Fixes #3961 Position of calculated column in data frame
@africanmathsinitiative/developers this is ready for review.

The changes in `dlgStringHandling `have only been made to demonstrate how the `ucrSave `control will be used in the dialogs.

### CODE REVIEW

**Comment by Patowhiz** [comment]
@africanmathsinitiative/developers this is now ready for review
For dialogs that will want the default position to be after the selected column without necessarily clicking on the position button , they will have to link the dialog receiver to the ucrSave control (
`ucrSaveStringHandling.setLinkedReceiver(ucrReceiverStringHandling)`  ) .

I have demonstrated this in the `dlgStringHandling`. The other dialogs currently will have the position being at the end until when the receiver is linked in them.

**Comment by Patowhiz** [comment]
@dannyparsons the only dialog that I found with an issue is the  `dlgRegularSequence `which is used in generating dates. After checking its code I've realised, the Rcode which is to be written by `ucrSave `control  is being updated manually. What's your advise on this in regard to the current implementation.

dlgregularsequence

**Comment by Patowhiz** [comment]
@maxwellfundi  could you help me in resizing & aligning the position button. My laptop resolution could be the problem here.

ucrsave

**Comment by Patowhiz** [comment]
This pull request affects several dialogs. Could someone check if it works well with all the dialogs. Keeping in mind that the only one linked right now is `dlgStringHandling`.

**Comment by Patowhiz** [comment]
@africanmathsinitiative/developers @dannyparsons this change is needed in the next release yet it will affect many dialogs . Testing this early and fixing the changes is a priority. I would appreciate if testing and reviews could be done this early.

**Comment by dannyparsons** [comment]
I agree this should be tested as a priority.

**Comment by Muthenya** [comment]
I have also noticed that when I click reset, the receiver in the subdialog doesn't reset.

**Comment by dannyparsons** [comment]
Someone else could realign once merged if needed. That will probably be quickest.

**Comment by Patowhiz** [comment]
> I have also noticed that when I click reset, the receiver in the subdialog doesn't reset.

@Muthenya in regard to your comment above, the subdialog does reset, but not in away that you expect. When you click on reset, the receiver becomes empty and the subdialog checks the End option and disables the receiver, since you can't have the column position after nothing.

**Comment by Muthenya** [comment]
This is what I mean.

capture

**Comment by Patowhiz** [comment]
@Muthenya can you test this again please. I have fixed the issue of reset.

**Comment by dannyparsons** [comment]
I think this shows that we would be better off hiding the selector and receiver for start and end. Otherwise it looks confusing to see a column selected.

**Comment by Muthenya** [comment]
This reset is working well. I also agree with Danny. It's better to hide them than have them disabled for start and end.

**Comment by Patowhiz** [comment]
Okay let me implement the suggestion

**Comment by Muthenya** [comment]
This is working well. I've also noticed the setting where if the user has checked the **Before** or **After** option and the column receiver in the subdialog is empty, then the check now goes to **Start** when it was previously on **Before** and **End** when it was previously on **After**, which is also great.

I'm not sure on the need for this but I was thinking about the possibility of having a message box appearing when the **Before** or **After** option is checked and the receiver is empty on clicking return.

**Comment by Patowhiz** [comment]
@dannyparsons what's your opinion on this. @Muthenya we had discussed this before with @rdstern and we had resolved to a functionality that you have seen above, since after and before nothing simply means end and start. I'm assuming you are suggesting a message box that notifies someone that his option has been changed to end or start ?

**Comment by Muthenya** [comment]
Exactly.

**Comment by dannyparsons** [comment]
It might be something to notify the user of, but then the position changes it other ways too e.g. resetting or changing the main receiver, so I think it's fine without a message box. We can see how it is used and if its something that would be useful to add.

**Comment by Patowhiz** [comment]
Okay, then this is ready for you @dannyparsons

File: instat/sdgSaveColumnPosition.vb [hunk 1]
```diff
@@ -91,8 +90,9 @@ Public Class sdgSaveColumnPosition
             InitialiseControl()
             bControlsNotInitialised = False
         End If
-        ucrSelectorColumns.Enabled = True
-        ucrReceiverColumn.Enabled = True
+        ucrSelectorColumns.Visible = True
+        ucrReceiverColumn.Visible = True
+        lblColumns.Visible = True
```
Comment (shadrackkibet) [inline instat/sdgSaveColumnPosition.vb]: Can all this be done by linking?

Comment (Patowhiz) [reply instat/sdgSaveColumnPosition.vb]: @shadrackkibet  Please elaborate on how this could be done by linking.

Comment (shadrackkibet) [reply instat/sdgSaveColumnPosition.vb]: will it be okay if i do the changes and explain what i have done?

Comment (Patowhiz) [reply instat/sdgSaveColumnPosition.vb]: Sounds good

**Comment by Patowhiz** [comment]
@dannyparsons this awaits your review. Based on @rdstern mail **R-Instat tasks & milestones in 2019**
on **Other improvements**, this seems urgent.

**Comment by dannyparsons** [comment]
Thanks for the reminder, yes I know this has been waiting for me to review for a while. I don't think I will get to it for the next release but I'll add it to the following milestone to remind me to review.

**Comment by maxwellfundi** [comment]
@dannyparsons having that we have no pressing milestone in the next few weeks, i know @rdstern would like this possibly for the next release. Could you review this soon?

**Comment by maxwellfundi** [comment]
@dannyparsons  this a quick reminder on review of this.

**Comment by dannyparsons** [comment]
Thanks, this is a good reminder. Tentatively, I hope we can complete this for the next release.

**Comment by maxwellfundi** [comment]
@dannyparsons  you now have some time for R-Instat work. Would you make sure this is in the list of the things you need to do soon

**Comment by rdstern** [comment]
This would be a good feature to resume work on

**Comment by maxwellfundi** [comment]
This requires Danny's review actually. I think Patrick is already done with
his bit of work

**Comment by Patowhiz** [comment]
Yes I was done with it long time ago. Though already there is conflicts, I had made changes to the core classes, which were also edited by Danny before merging.

Once merged this PR will enhance the dialogs which create new columns.

**Comment by dannyparsons** [comment]
Can you easily resolve the conflicts? That needs to be done before we can look to merge.

**Comment by Patowhiz** [comment]
Yes, I'll consult @maxwellfundi in doing this.

**Comment by maxwellfundi** [comment]
@Patowhiz  Let me  have a look at the conflict and if its easy to fix then we can fix it

**Comment by Patowhiz** [comment]
I have resolved the conflicts and tested the changes. @africanmathsinitiative/developers this is now ready for review

**Comment by rdstern** [comment]
@Patowhiz I have tried it there and like it a lot.
I usually tried with new names and they do as you would expect.
I then tried with the same name as before and it didn't move the variable, just changed the name.  I suggest that is fine.

So, it seems to work fine.
A small detail is the actual control is not lined up exactly with the new column name.  I think it is also a few points deeper, but that may be my imagination.

Perhaps the settings of the control might be remembered for new dialogues.  I can't tell that. 
 Otherwise what is the initial default?

Could we now try in a dialogue like **_Prepare > Column: Calculate > Calculations_** where there may be a variable number of columns used.  Perhaps also in **_Prepare > Dates > Make Date_**, where the different tabs allow for one or two or 3 columns to make the new variable.  Of course also in Prepare > Data Frame > Duplicate, so that dialogue can then also (later) be changed.

**Comment by maxwellfundi** [comment]
great @Patowhiz .

**Comment by maxwellfundi** [comment]
@Patowhiz  @rdstern  I have been testing this and I think I have a few issues that dont seem to work right.  I use the regular sequence dialog to generate a regular sequence 1- 36 using experimental data

1. I go to the dialog and do the default, and my column Regular1 comes at the end. Thats works. 

image

2. I now decide i want my column to be at the start, I click on the position and select start- The column does not appear in the correct position- It goes at the end again (I thought this would have been the first column)

image

3. I then want my column position to be before size, I click on the position and select before, then select size- On clicking ok the column comes again at the end 

image

4. Finally, second thing that seemed to work right, I decide i want it after field and filled in my dialog as shown 

image


Then on clicking okay, it comes two columns away from my intended position
image


@rdstern  I dont know if you encountered these issues.   

@Patowhiz  I think you need to check if the parameters are being passed properly for the column positions. 

While we were fixing the conflicts, could you also check that we kept all the changes you needed here. maybe something changed or was left out.

**Comment by rdstern** [comment]
Aha @maxwellfundi I had only tested on the strings, where a single variable was  being transformed.  I am getting your problem on the regular sequence dialogue.  
There we are producing a new variable, but there isn't an original variable we are producing it from.  
I tried with the Dodoma data and the calculated variable usually comes at the end.  When I asked for the Start it came second, and just after the variable with the focus in the sheet.
I also have a request for situations like this - perhaps we made it before?  When (as in the regular sequence dialogue) there isn't an obvious variable to be before (or after), could it check the focus, i.e. where we are in that dialogue.  Then it completes the sub-dialogue with that element.

This is like the right-click on say rename, where it picks out the name.  There is a difference, because we could use regular sequence on a different data frame to the one that is visible.  I don't know if the variable on the current focus is available then.  If not, then it does what it can.

**Comment by Patowhiz** [comment]
@maxwellfundi @rdstern I request both of you to kindly check on the conversation on this PR from the start. This is a problem that both @dannyparsons and I are already aware of. I did check on the regular sequence dialog and reported on my findings(check above in this conversation thread). 

I'm still interested in knowing about the other dialogs, there could be another one implemented like the regular sequence, we need to identify them.

@rdstern I checked on the `Make date`s dialog, it works correctly on my end, the `Calculate `dialog uses a different control, I think to use this feature we will have to first have it use this control `(ucrSave)`

**Comment by rdstern** [comment]
I have now read through the whole thread as @Patowhiz suggested.  I can see why he was getting incensed that we are in danger of "going round in very old circles".  
I also discovered that he has implemented the new control in a great many dialogues.
And he has asked repeatedly for help from someone else on resizing the new position button.  Can someone help now, or should we merge and sort it out later.

I suggest either @dannyparsons looks quickly and merges, or @maxwellfundi merges and we continue from there.  

If we do merge  then could @Patowhiz (or someone else) work quickly on the new design for the duplicate dialogue.  That needs to have the old column position taken out and the type of column added in.  I think @Patowhiz has something just-about ready to implement on this.

I would still like a further feature added - and didn't see it much on the thread - where an initial example is the Rank or Count in Factor, etc.  These are dialogues, where a single variable is producing a new variable.  Then the default is that **_before and after_** is in relation to this single variable.  So that is inserted automatically when possible.

I would like to consider other small improvements as well.  But let's get it merged first?

**Comment by maxwellfundi** [comment]
@Patowhiz  I think you are right. I suggest if we can use this control on the dialogs that produce columns then we can

**Comment by Patowhiz** [comment]
@rdstern all the dialogs that have this control can now set the default column position quite easily. They just need to link the receiver to the `ucrSave `and everything else gets handled automatically(We need the receiver linking because it's the only way we can determine what's currently selected). I implemented this based on the suggestion you made on the issue about the default column position to be after the selected column/variable.

I have demonstrated this in the String handling dialog
`ucrSaveStringHandling.setLinkedReceiver(ucrReceiverStringHandling)`

After this is merged we can identify the dialogs which we want this to happen by default and add just a single line of code to them. That means dialogs like **Rank** and **Count in factor** can benefit from this feature.

**Comment by rdstern** [comment]
Great. Thanks.

**Comment by dannyparsons** [comment]
I suggest not adding further work on this PR and making new issues for further features once this is merged. I will take this to go through this now in hope to get it merged this week.

**Comment by rdstern** [comment]
@dannyparsons can this be merged now, as @Patowhiz is now here and we could move to complete this feature?

**Comment by Patowhiz** [comment]
@dannyparsons could you please finalise this. I have an incoming PR for the receivers that will also affect several dialogs. It will be good if this was checked and adopted to the other dialogs before we merge the one for the receiver

**Comment by maxwellfundi** [comment]
@Patowhiz I know you have issues with the sizing of designers on your computer.  @Wycklife  could you help @Patowhiz  resize this?

**Comment by rdstern** [comment]
Yes, but @volloholic looked through the settings/code and made a suggestion that just could easily resolve the issue.  And if not, then let's merge now, so we get on.

**Comment by volloholic** [comment]
> @Patowhiz I know you have issues with the sizing of designers on your computer. @Wycklife could you help @Patowhiz resize this?

@Patowhiz and @maxwellfundi This is not primarily a sizing issue it is a settings issue related to using auto size for the button.

**Comment by maxwellfundi** [comment]
okay great @volloholic. When that is fixed I will get it merged.

**Comment by Patowhiz** [comment]
@maxwellfundi I understand what you mean. Visual studio 2017 doesn't have the feature to set scaling, therefore devices that had very high resolutions were affecting form designs for the winforms. This meant I needed help with any form or control design for winforms. But now I'm using Visual studio 2019, which has this feature.

@volloholic I made the change that you requested but it still doesn't align well, after checking this closely, all the other controls in this control are custom controls. Therefore, the controls inside the control also have their own margins. The button position is not a custom control that's why it looks misaligned. Basically, it's a problem of the differences between custom control margins and none custom controls. 

I think having a custom button control will help with the margins since it will behave like the other custom controls.

**Comment by volloholic** [comment]
@Patowhiz There are other ways to fix this however @rdstern and I have been discussing the need for a custom button in the calculators where buttons manage rcode. I think there is an interesting case here where we might want to assign the code from the subdialog to the button for the position. This probably needs more discussion when we are together, so you could either merge as is or get started on implementing something and we will discuss next week.

**Comment by maxwellfundi** [comment]
I will then get this merged then, further work can start and be on a new PR.

**Pull request merged**